### PR TITLE
Enable service ca bundle injection for webhooks

### PR DESCRIPTION
**Pull request by marun**
/cc @deads2k @sttts @stlaz

### CODE REVIEW

File: enhancements/service-ca-bundle-injection-for-webhooks.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Webhooks can secure their endpoints automatically with a serving cert
```
Comment (sttts) [inline enhancements/service-ca-bundle-injection-for-webhooks.md]: Be more precise throughout the doc that this is about admission webhooks. We have a number of webhooks in kube meanwhile.

File: enhancements/service-ca-bundle-injection-for-webhooks.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+simple and well-tested option.
+
+### Goals
+
+- Service CA bundle injection is supported for both mutating and
+  validating webhook configurations.
+
+### Non-Goals
+
+- Supporting ca bundle injection to a subset of webhooks defined in a
```
Comment (sttts) [inline enhancements/service-ca-bundle-injection-for-webhooks.md]: nit: CA everywhere

File: enhancements/service-ca-bundle-injection-for-webhooks.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  webhook configuration resource.
+  - Allowing selective injection would likely increase the complexity
+    of implementation and there is no clear indication that this
+    capability is required.
```
Comment (sttts) [inline enhancements/service-ca-bundle-injection-for-webhooks.md]: ... Webhooks in one configuration object are all independent and therefore configuration can be split into multiple resources if difference CAs are necessary.

File: enhancements/service-ca-bundle-injection-for-webhooks.md (hunk 1 above)
Comment (marun) [reply enhancements/service-ca-bundle-injection-for-webhooks.md]: Done

File: enhancements/service-ca-bundle-injection-for-webhooks.md (hunk 2 above)
Comment (marun) [reply enhancements/service-ca-bundle-injection-for-webhooks.md]: Done

File: enhancements/service-ca-bundle-injection-for-webhooks.md (hunk 3 above)
Comment (marun) [reply enhancements/service-ca-bundle-injection-for-webhooks.md]: Done

**Comment by marun** [comment]
@sttts Updated, PTAL

**Comment by sttts** [comment]
/lgtm
/approve

/hold

for @deads2k to take a final look

**Comment by marun** [comment]
Moved the new enhancement to the `authentication` path.

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
+---
+title: service-ca-bundle-injection-for-admission-webhooks
```
Comment (deads2k) [inline enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: make the enhancement for injection in general

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+- Supporting CA bundle injection to a subset of webhooks defined in an
+  admission webhook configuration resource.
+  - Allowing selective injection would likely increase the complexity
+    of implementation and there is no clear indication that this
+    capability is required.
+  - Webhooks in one configuration object are all independent and
+    therefore configuration can be split into multiple resources if
+    difference CAs are necessary.
+
+## Proposal
```
Comment (deads2k) [inline enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: list out the other spots where we make use of this please and their behavior so we can compare.  Not all of us remember.  I think I recall a configmap and an apiservice.

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+- Add a new bundle injection controller for `MutatingWebhookConfiguration`
+- Add a new bundle injection controller for `ValidatingWebhookConfiguration`
```
Comment (deads2k) [inline enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: add an example of each and highlight how if you embed multiple webhooks in one configuration resource they will *all* get injected.

**Comment by deads2k** [comment]
Why would I do this here in preference to an upstream CSR signer that has a trust bundle wired up to be injected or auto-trusted via an option?

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+- The new controllers will ensure that both types of admission webhook
+  configurations will have all their CABundle fields populated by the
+  current service CA bundle when they are found to have one of the
+  injection annotations (`service.beta.openshift.io/inject-cabundle`
+  or `service.alpha.openshift.io/inject-cabundle`)
```
Comment (sttts) [inline enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: as @deads2k pointed out, we should probably stop supporting the alpha annotation for any new resource (in 4.4 and beyond).

**Comment by sttts** [comment]
@marun please address the open questions and suggestions in a follow-up. I will unblock this now to get merged before freeze as the outcome of the enhancement is clear.

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md (hunk 7 above)
Comment (marun) [reply enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: PR implementing your request: https://github.com/openshift/service-ca-operator/pull/102

Comment (sttts) [reply enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: can you start the update of the enhancement?

Comment (marun) [reply enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: In progress

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md (hunk 4 above)
Comment (marun) [reply enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: Done

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md (hunk 5 above)
Comment (marun) [reply enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: Done

**Comment by marun** [comment]
> Why would I do this here in preference to an upstream CSR signer that has a trust bundle wired up to be injected or auto-trusted via an option?
> 
> This is a question about where we're going and how we'll transition. It's not a "no" to this feature.

@deads2k Is it intended that in the future CSR signing will replace the function provided by the service ca operator? If so, has a timeline been discussed as to the transition?

File: enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md (hunk 6 above)
Comment (marun) [reply enhancements/authentication/service-ca-bundle-injection-for-admission-webhooks.md]: Done

**Comment by sttts** [comment]
> @deads2k Is it intended that in the future CSR signing will replace the function provided by the service ca operator? If so, has any planning been done as to strategy and timeline?

I don't think there has been much discussion upstream. But the topic comes up again and again that cert generation ought to be part of the platform, e.g. https://twitter.com/ibuildthecloud/status/1220798262187159552.

**Pull request merged**